Fps Profiler - save fps, cpu, gpu data to csv file

w-czerski opened the pull request:
## About
This PR introduces a game mode fps and memory statistics profiler that exports collected data into a `csv` file.

## Features
The Profiler has an EBus that can control profiling in runtime (start/stop/reset), save profiled data, change save path, and access current frame memory data or fps (avg, min, max). Works with `Lua` and `Script Canvas`.

> _For details please refer to readme._

## Saved File Example
| Frame | FrameTime | CurrentFPS | MinFPS | MaxFPS | AvgFPS | CpuMemoryUsed | GpuMemoryUsed |
|-------|----------|------------|--------|--------|--------|---------------|--------------|
| 1     | 0.1731 | 0.00       | 2.38| 0.00  | 0.00  | 1349.42       | 1752.12      |
| 2     | 0.4207    | 2.38       | 2.38   | 2.38   | 1.19   | 1375.50       | 2999.38      |
| 3     | 0.1934    | 5.17       | 2.38   | 5.17   | 2.52   | 1400.49       | 2963.44      |

## Editor Configuration
![Screenshot from 2025-03-17 15-31-39](https://github.com/user-attachments/assets/8881984b-e0ad-4187-9858-c96b1fee0b5d)


## Possible Improvements
This can be extended to profile the editor data, for example, with exposed EBus functionality.

w-czerski committed:
init

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
Add profiler configuration

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename fix

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove unused

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove unused services

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix headers

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
debug display on screen

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add siplay fps on screen | add refleciton of config

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
improvements

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clean up

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix game launcher ddebug display

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
save with timestamp

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
change fps color

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
Add notification bus for created files

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add warning

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix warning - use c_str

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
change name

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add info in tick
:

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
refactor

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix calcuatlion of avergage frame

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix precision to near zero values

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add more editor options | specify precision and occurrance size

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add cpu and gpu memory collect

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clean up tick | clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
make memory getter functions static

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
use detailed memory flag

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix fps data struct reflection | fix system component, clean up, adjust to config

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
refactor

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
refactor

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
refactor | remove redundant headers | remove rpi from cmake, use only rhi public

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename to Fps | use const reference for notification bus

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
move static functions to public

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix near zero editor reflection

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add request bus info | implement bus overrides

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix activate / deactivate

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename profiler data -> profiler config

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add fix config layout

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add fix config layout

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add is any save option enabled

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add option to change save path

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
typo fix

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
exit early when no save option enabled

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
mark disabled data to -1.0f | using 0.0f might be confusing

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix near zero comparison

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add option to disable instant profiling in editor

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix profiling setup on tick and activate

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -0,0 +1,401 @@
+        return m_configuration.m_SaveFpsData || m_configuration.m_SaveCpuData || m_configuration.m_SaveGpuData;
+    }
+
+    void FPSProfilerSystemComponent::ChangeSavePath(const AZStd::string& newSavePath)
+    {
+        AZ_Printf("FPS Profiler", "Path changed.");
+        m_configuration.m_OutputFilename = newSavePath;
+    }
```
I'm not exactly convinced if allowing user to change log filename during profiling is a good thing. But if for some reason it would be useful, maybe it is good idea to warn them that log filename changed during profiling. I would also consider some sort of validation if path given via string is a valid

w-czerski replied:
I've assumed user might specify different directories without multiple times running the simulation. 
User needs to use it with extreme caution, this is why I've introduced SafeChangePath.
I've added extra caution log info.

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -0,0 +1,401 @@
+    void FPSProfilerSystemComponent::CreateLogFile()
+    {
+        AZ::IO::FileIOBase* fileIO = AZ::IO::FileIOBase::GetInstance();
+        bool fileExists = fileIO->Exists(m_configuration.m_OutputFilename.c_str());
+
+        if (!fileExists)
+        {
```
But if someone put custom path e.g via bus, this file doesn't exists. In such case forcing it to default value seems suspicious. You can consider file creation in such case, of course only if given path is valid. Only in case of invalid path, default shall be restored

w-czerski replied:
Done

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -0,0 +1,401 @@
+
+    void FPSProfilerSystemComponent::WriteDataToFile()
+    {
+        // Exit when nothing to save
+        if (m_logEntries.empty())
+        {
+            return;
+        }
+
+        // If none save option enabled - exit
+        if (!IsAnySaveOptionEnabled())
+        {
+            return;
+        }
```
Nit: this code is self explanatory, i'm not sure if those comments add something new

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -0,0 +1,401 @@
+        if (!fileExists)
+        {
+            m_configuration.m_OutputFilename = "@user@/fps_log.csv"; // Restore to default path
+            AZ_Error(
+                "FPSProfiler::CreateLogFile",
+                false,
+                "Specified file does not exist. Using default path: %s",
+                m_configuration.m_OutputFilename.c_str());
```
Nit: This is rather warning than error

norbertprokopiuk reviewed (commented):
Feature works as expected. I left some comments for code improvement

w-czerski commented on Gems/FPSProfiler/Code/Source/Tools/FPSProfilerEditorSystemComponent.cpp:
```diff
@@ -0,0 +1,62 @@
+
+        if (auto serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serializeContext->Class<FPSProfilerEditorSystemComponent>()
+                ->Version(0)
+                ->Field("m_Configuration", &FPSProfilerEditorSystemComponent::m_configuration)
+                ->Field("m_profileOnGameStart", &FPSProfilerEditorSystemComponent::m_profileOnGameStart);
```
Got it, haven't received that error before.

w-czerski commented on Gems/FPSProfiler/Code/Include/FPSProfiler/FPSProfilerBus.h:
```diff
@@ -0,0 +1,85 @@
+        virtual size_t GetGpuMemoryUsed() const = 0;
+
+        // Logging
+        virtual void SaveLogToFile() = 0;
+        virtual void SaveLogToFile(const AZStd::string& newSavePath, bool useSafeChangePath = true) = 0;
```
Changed

w-czerski commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -0,0 +1,401 @@
+        }
+
+        // If none, dont proceed
+        if (!(m_isProfiling && m_configuration.m_ShowFps))
```
Right, I've removed early exits.

w-czerski committed:
apply code review | add path validation | code improvement

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang tidy & format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
valdiate path on activation

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
update reflect name

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix file creation | timestamp

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -302,22 +300,39 @@ namespace FPSProfiler
+        {
+            m_configuration.m_OutputFilename = "@user@/fps_log.csv";
+            AZ_Warning("FPSProfiler", false, "Invalid output file path. Using default: %s", m_configuration.m_OutputFilename.c_str());
+            return;
```
This return prevents log creation in case of providing invalid path, even after restoring it to default value

w-czerski replied:
Resolved

norbertprokopiuk commented on Gems/FPSProfiler/Code/Include/FPSProfiler/FPSProfilerBus.h:
```diff
@@ -18,24 +19,26 @@ namespace FPSProfiler
         // Logging
         virtual void SaveLogToFile() = 0;
-        virtual void SaveLogToFile(const AZStd::string& newSavePath, bool useSafeChangePath = true) = 0;
+        virtual void SaveLogToFileWithNewPath(const AZStd::string& newSavePath, bool useSafeChangePath) = 0;
```
What is the reason to change path to `AZ::IO::Path` in `ChangePath` methods, but keeping in here as string? Could it be consisted across the API?

w-czerski replied:
Thanks for noticing, changed all to `AZ:IO:PATH`.

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -201,13 +194,18 @@ namespace FPSProfiler
+        }
+
         m_configuration.m_OutputFilename = newSavePath;
+        AZ_Warning("FPS Profiler", false, "Path changed.");
```
```suggestion
        AZ_Warning("FPS Profiler", !m_isProfiling, "Path changed during activated profiling.");
```
Warning about changed path should be triggered only when there is any danger associated with it. Also warning message should include information about the danger

w-czerski replied:
Fair point - updated code.

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -302,22 +300,39 @@ namespace FPSProfiler
+            }
         }
 
         if (m_configuration.m_SaveWithTimestamp)
```
This check should be before attempting to check if path exists. You change filename, so even if you checked and created file, it doesn't matter because you potentially changed filename in this if statement.

w-czerski replied:
File name is altered before file creation.

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -302,22 +300,39 @@ namespace FPSProfiler
+                AZ_Printf("FPSProfiler", "Log file successfully created: %s", m_configuration.m_OutputFilename.c_str());
+            }
+            else
+            {
+                // Restore default path on fail
+                m_configuration.m_OutputFilename = "@user@/fps_log.csv";
+                AZ_Warning("FPSProfiler", false, "Failed to create file. Using default path: %s", m_configuration.m_OutputFilename.c_str());
+            }
```
you already validated given path and set correct value. What is the reason to potentially change it again?

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -302,22 +300,39 @@ namespace FPSProfiler
+            AZ_Warning("FPSProfiler", false, "File does not exist, trying to create it...");
+
+            AZ::IO::HandleType fileHandle;
+            if (fileIO->Open(
+                    m_configuration.m_OutputFilename.c_str(), AZ::IO::OpenMode::ModeWrite | AZ::IO::OpenMode::ModeCreatePath, fileHandle) ==
+                AZ::IO::ResultCode::Success)
+            {
+                fileIO->Close(fileHandle);
+                AZ_Printf("FPSProfiler", "Log file successfully created: %s", m_configuration.m_OutputFilename.c_str());
+            }
```
There is a method to create file in fileIO. Something like this should do the job. But you should decide on one method of file creation, because right now you use two. One here and the other one in line 357 in `AZ::IO::FileIOStream` contructor using flag `AZ::IO::OpenMode::ModeCreatePath`. In my opinion it would be beneficial to keep the flow like so:
1. check if path is valid, assign new value if path invalid
2. check if timestamp needed // path still valid since only filename changes
3. create/open file, write data and close file. // lines 357-360 do this point
This is more or less what I meant in my previous comment related to this code. Previous solution was mostly correct. My only concern was about path validation and assigning new value in case of invalid value. Sorry if my previous comment was misleading.
```cpp
auto result = fileIO->CreatePath(m_configuration.m_OutputFilename.c_str());
```

w-czerski replied:
Got it. I was checking is it possibly to open file. I've removed redundant code and simplified it.

w-czerski committed:
fix memory reserve

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
replace string with IO:PATH

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove redundant code

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
change string path to AZ:IO:PATH

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
use deque instead of vector

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
reflect missing variable

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add info to gem.json

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk reviewed (commented):
Please add section describing feature in repo readme to follow the standard introduced by other gems from this repo

w-czerski committed:
init readme

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add context to readme

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add extra notification buses for stop/start/reset

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk reviewed (changes requested):
I did some CPU time profiling and I found that this Gem takes a lot of time to calculate its data and causes significant frame rate drop. On my machine it was roughly 20FPS in editor (60->~40). Due to that this Gem doesn't provide reliable information even though it works as intended. I managed to track down part of code causing the issue, which needs to be resolved before merging.

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -0,0 +1,424 @@
+            return;
+        }
+
+        AZStd::string logEntry = AZStd::string::format(
+            "%d,%.4f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f\n",
+            m_configuration.m_SaveFpsData ? m_frameCount : -1,
+            m_configuration.m_SaveFpsData ? deltaTime : -1.0f,
+            m_configuration.m_SaveFpsData ? m_currentFps : -1.0f,
+            m_configuration.m_SaveFpsData ? m_minFps : -1.0f,
+            m_configuration.m_SaveFpsData ? m_maxFps : -1.0f,
+            m_configuration.m_SaveFpsData ? m_avgFps : -1.0f,
+            m_configuration.m_SaveCpuData ? BytesToMB(GetCpuMemoryUsed()) : -1.0f,
+            m_configuration.m_SaveGpuData ? BytesToMB(GetGpuMemoryUsed()) : -1.0f);
+        m_logEntries.push_back(logEntry);
+
```
part of code related to requested changes and performance drop

w-czerski replied:
Thanks for pointing out. I'm checking it right away.

w-czerski committed:
add print reset info

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
restore version bump

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerSystemComponent.cpp:
```diff
@@ -384,7 +365,6 @@ namespace FPSProfiler
         }
         file.Close();
-        m_logEntries.clear();
         m_fpsSamples.clear();
```
Indeed

w-czerski committed:
clear log entires fix

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix comments for deque

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
init editor variables

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
update comment variable names

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove string format | use char vector | fix buffer allocation

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove redundant comments

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
use Basic report flag | save 5.0ms

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove comment

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove redundant code

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clean up

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add biref summary for profiler bus

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add to log buffer reserved memory data

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
change sizte_to -> AZStd::size_t

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix layout | move buffer size to constexpr outside scope

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix table in readme | imrpove readability

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix readme table

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski commented:
I've managed to reduce complexity OnTick() to around **0.1ms**.
![Screenshot from 2025-03-04 16-59-20](https://github.com/user-attachments/assets/f44cc464-da1c-4b18-8c52-e27ef4afccdd)

w-czerski committed:
restore affected files by clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
use max buffer line size for memory reserve

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove if else | improve readability

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove profile flag

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add warning print for stop/reset

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
queue start profile on activation | add aditional bool flag

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
explanation why use minFps = FloatMax

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clean up | remove profiling flag

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
refactor layout of config

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename config file

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
move to config structs to namespace Config

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename config | add multiple config strutcs | add type ids

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix mask bit shifts

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
extra option

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
create reflect for RecrodSettings

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add other config refelcetions

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add new config to system component

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove enum int cast

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
replace old config | build success

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
change version to 0 | use static cast

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove reflect duplicates

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix tick comparison

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove bool from enum

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add EMA calculation | add smoothing factor

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
refresh ui

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
enable print when debug selected

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
reflect enum

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove bool toogles | using enum values

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix mask comparison

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
typo fix

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix save data on exit

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix prfoile start

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
deactivate fix

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
simplyfy frame count

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove redundant if statement4

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
apply seconds to timestamp

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix comment

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix file save on deactivation

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove enum serialization

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
Add Config reflection to System Component

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
create config dir | move config | fix config reflect

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
simplify if check

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add final to system component

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
swap IO Path to String | add button to add path

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix headers

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename function

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove system in name suffix

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
use path.c_str

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename System Component -> Component | remove unused

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename type ids | fix uuid duplicate

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
resolve serialize duplicates

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix path selector and validation

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add lua and script canvas support

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix path swap to string | operate on IO::Path

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
option to keep hisotry for better precision

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix keep history setting | fix reflect description

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
simplify logic | clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix game launcher | pass settings from editor

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix formatting | make fields protected

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add button description

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add better description

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
refactor

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix config reflect

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add debug visiblity on enabled button

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
readme update

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
readme table update

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
update images resolution

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add behaviour context for configs | update profile bus

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
readme update | fix behaviour reflect

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk reviewed (commented):
Note for future: Avoid merging main/development back to your feature branch, since it creates strange structure in git graph and shows some of already reviewed changes from main branch as changes currently made. Consider usage of the  git rebase instead

Overal work really nice and provides very useful data, great job! Although I left some comments related to design and one small regression with file paths

I would recommend removing Editor component since it doesn't have any editor specific operations, just reflection. Instead reflection for edit context can be added to game component (FPSProfileComponent). 

Could you also explain definitions of the system components in this setup? Possibly other than defining interfaces? Because this is the only reason that i see to define system components in this setup. But you defined all of them as system and don't really know why. It would be great if you could clarify this

w-czerski committed:
fix formatting

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix formatting cmake

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add missing new line

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
ediotr context for component

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove editor source files from cmake

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove ediotr component

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove editor component depencencies

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove editor component type id

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
change path string into AZ::IO::Path

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove unecessary checks for config refelections

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
make config private | refactor

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove const from bus notifies - blocking file modifications

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fixed fps counter display positioning

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski commented:
- [x] Replaced all `AZStd::string` path into `AZ::IO::Path`.
- [x] Removed file picker dialog.
- [x] Fixed path selection. 
- [x] Removed the editor component and system component, editor reflection is passed dawn by FPSComponent.
- [x] Removed duplicate reflection validation
- [x] Removed const from bus notifications (let user modify configs on the run if needed)
- [x] Fixed FPS counter visibility - ImGui toolbar is no longer covering the counter

w-czerski committed:
fix readme conflicts

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
readme update

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
Merge branch 'main' into wc/fps_profiler

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Clients/FPSProfilerComponent.cpp:
```diff
@@ -516,19 +530,17 @@ namespace FPSProfiler
+                : !path.HasFilename()         ? "Path must have a file at the end."
+                : !path.HasExtension()        ? "Path must have a *.csv extension."
+                : !fileIO                     ? "Could not get a FileIO object. Try again."
+                                              : "Path is not registered or recognizable by O3DE FileIO System.";
```
this nested if cannot stay here, because you checks the same thing multiple times. If you want to keep the logic with adding reason to warning, you can change `IsPathValid` to use `AZ::Outcome<bool,AZStd::string>`. Then your function would look something like this:
```cpp

    AZ::Outcome<bool, AZStd::string> FPSProfilerComponent::IsPathValid(const AZ::IO::Path& path) const
    {
        AZ::IO::FileIOBase* fileIO = AZ::IO::FileIOBase::GetInstance();

        if (!fileIO)
        {
            return AZ::Failure("Could not get a FileIO object. Try again.");
        }
        else if (!fileIO->ResolvePath(path))
        {
            return AZ::Failure("Cannot resolve the path");
        }
        else if (path.empty())
        {
            return AZ::Failure("Path cannot be empty");
        }
        else if (!path.HasFilename())
        {
            return AZ::Failure("Path must have a file at the end.");
        }
        else if (!path.HasExtension())
        {
            return AZ::Failure("Path must have a *.csv extension.");
        }
        else
        {
            return AZ::Success(true);
        }
    }
``` 

and then check `if(!IsPathValid)` would change to something like this:
```cpp
        auto pathValidationOutcome = IsPathValid(m_configFile.m_OutputFilename);
        if (!pathValidationOutcome.IsSuccess())
        {
            AZ_Warning("FPSProfilerComponent::CreateLogFile", !m_configDebug.m_PrintDebugInfo, pathValidationOutcome.GetError().c_str());
         // further implementation
         }
```

w-czerski replied:
Got it, updated.

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Configurations/FPSProfilerConfig.cpp:
```cpp
      ->DataElement(
                        AZ::Edit::UIHandlers::Default,
                        AZ::Edit::UIHandlers::Default,
                        &FileSaveSettings::m_OutputFilename,
                        &FileSaveSettings::m_OutputFilename,
                        "Csv Save Path",
                        "Csv Save Path",
                        "Select a path where *.csv will be saved.")
                        "Select a path where *.csv will be saved.")
                    ->Attribute(AZ::Edit::Attributes::SourceAssetFilterPattern, "*.csv")
                    ->Attribute(AZ::Edit::Attributes::SourceAssetFilterPattern, "*.csv")
```
This part of reflection still doesn't work as it is supposed to. Even after selecting path via Editor, log is saved at project_root/Cache/linux/

I would suggest to for now remove the possibility to change path from edit context, and create issue with low priority to handle this. It is really minor thing and it is not worth to keep this PR on hold because of that.

w-czerski replied:
I've makred for now this editor property as read only

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/FPSProfilerModuleInterface.cpp:
```diff
@@ -24,11 +24,4 @@ namespace FPSProfiler
-        return AZ::ComponentTypeList{
-            azrtti_typeid<FPSProfilerComponent>(),
-        };
-    }
```
I wouldn't remove it, I would rather pass empty list. It is override after all

norbertprokopiuk commented on Gems/FPSProfiler/Code/Source/Tools/FPSProfilerEditorModule.cpp:
```diff
@@ -20,20 +20,9 @@ namespace FPSProfiler
-            return AZ::ComponentTypeList{
-                azrtti_typeid<FPSProfilerEditorComponent>(),
-            };
-        }
```
same as form previous module.cpp file

norbertprokopiuk reviewed (commented):
Looks really good, I left only some comments to improve minor things. 

But I will point it once again. Try to avoid merging main branch to your feature branch. It really messes up git graph structure. Better practice in case of conflicts is to update your local repo to be up to date with the remote, pull latest commits from the main branch. git switch/checkout to your feature branch. Git rebase main and then resolve all conflicts on you local repo. When you arre done, push rebased changes to the feature branch by git push origin feature_branch --force-with-lease.

w-czerski committed:
remove pure vritual from notify bus

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add GetRequiredComponentType list override

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
mark path seleciton as read only

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
update path validation function

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski commented on Gems/FPSProfiler/Code/Include/FPSProfiler/FPSProfilerBus.h:
Reverted to previous set up

norbertprokopiuk reviewed (approved):
Good job, very nice feature!
